nixos/modules/virtualisation/keystone.nix

```diff
@@ -0,0 +1,211 @@
+        '';
+      };
+
+      adminUsername = mkOption {
+        type = types.str;
+        default = "admin";
+        description = ''
+          A keystone admin username.
+        '';
+      };
+
+      adminPassword = mkOption {
+        type = types.str;
+        default = "admin";
+        description = ''
+          The keystone admin user's password.
+        '';
+      };
+
+      adminTenant = mkOption {
+        type = types.str;
+        default = "admin";
+        description = ''
+	  A keystone admin tenant name.
+        '';
+      };
+    };
+
+    database = {
+      host = mkOption {
+          type = types.str;
+          default = "localhost";
+          description = ''
+            Host of the database.
+          '';
+      };
+      name = mkOption {
+        type = types.str;
+        default = "keystone";
+        description = ''
+          Name of the existing database.
+        '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "keystone";
+        description = ''
+          The database user. The user must exist and has access to
+          the specified database.
+        '';
+      };
+
+      password = mkOption {
+        type = types.str;
+        default = "keystone";
+        description = ''
+          The database user's password.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    # Note: when changing the default, make it conditional on
+    # ‘system.stateVersion’ to maintain compatibility with existing
+    # systems!
+    virtualisation.keystone.package = mkDefault pkgs.keystone;
+
+    services.mysql.enable = mkDefault (cfg.database.host == "localhost");
+    services.mysql.package = mkDefault pkgs.mysql;
+
+    users.extraUsers = [{
+      name = "keystone";
+      group = "keystone";
+      uid = config.ids.uids.keystone;
+    }];
+    users.extraGroups = [{
+      name = "keystone";
+      gid = config.ids.gids.keystone;
+    }];
+
+    systemd.services.keystone-all = {
+        description = "OpenStack Keystone Daemon";
+        after = [ (optionalString (cfg.database.host == "localhost") "mysql.service") "network.target"];
+        requires = [ (optionalString (cfg.database.host == "localhost") "mysql.service") ];
+        path = [ cfg.package pkgs.mysql pkgs.curl pkgs.pythonPackages.keystoneclient pkgs.gawk ];
+        wantedBy = [ "multi-user.target" ];
+        preStart = ''
+          mkdir -m 755 -p /var/lib/keystone
+          # Initialise the database
+          ${cfg.package}/bin/keystone-manage --config-file=${keystoneConf} db_sync
+          # Set up the keystone's PKI infrastructure
+          ${cfg.package}/bin/keystone-manage --config-file=${keystoneConf} pki_setup --keystone-user keystone --keystone-group keystone
+        '';
+	postStart = optionalString cfg.bootstrap.enable ''
+	    # Wait until the keystone is available for use
+            count=0
+            while ! curl -s  http://localhost:35357/v2.0 > /dev/null 
```

Comment:
Probably the port could be configurable (actually I'm not sure actually if it is related to endpointPublic or not). There was no other parameter to specify listening address.